File: datamigrator/studio/templates.py
"""Starter templates for the Studio sidebar: pick one, fill a small form, get a
working mapping / chain / plan (or start a run) instead of building it from an
empty canvas.

Each entry in CATALOG is pure metadata — the Studio renders its form from the
`params` schema — and APPLY maps the template's slug to the function that
actually creates the objects. Everything a template creates is an ordinary
Mapping / CallChain / MigrationPlan, editable and deletable like any other;
nothing here is a special object type. Templates that only start something
(`action`) have no server side: the Studio opens the mapping and its run dialog.
"""
import json
import re

from django.db import transaction

from chains.models import CallChain, CallChainStep
from connections.models import Connection
from mappings.models import EntityMapping, FieldMapping, Mapping
from plans.models import MigrationPlan, PlanStep
from schemas.models import Entity, Field


class TemplateError(ValueError):
    """A problem with the user's answers — its message is shown as-is."""


# ── Catalog (metadata the Studio renders) ────────────────────────────────────

def _p(name, label, type="text", **kw):
    return {"name": name, "label": label, "type": type, **kw}


_NAME = _p("name", "Name", required=True, placeholder="e.g. Products → ERP")
_METHOD = _p("write_method", "Operation", "select", options=["POST", "PUT", "PATCH"], default="POST",
             help="HTTP verb used to write each record to the destination.")
_TRANSFORM = _p("transform", "Clean up text fields", "select", default="none", options=[
    {"value": "none", "label": "Leave values as they are"},
    {"value": "trim", "label": "Trim whitespace"},
    {"value": "uppercase", "label": "UPPERCASE"},
    {"value": "lowercase", "label": "lowercase"},
], help="Added as a transform on every text field that is copied to a text field. Editable per wire afterwards.")
_CONNECTION = _p("connection", "Connection", "connection", required=True, help="Every step calls this connection's API.")
_CHAIN_NAME = _p("name", "Name", required=True, placeholder="e.g. Create customer + order")

CATALOG = [
    # ── migrations (mappings) ──
    {"slug": "copy-matching-fields", "kind": "mapping", "icon": "bi-magic", "title": "Copy matching fields",
     "blurb": "Pick a source and a destination entity — every field with the same name is wired for you.",
     "params": [_NAME,
                _p("source_connection", "Source connection", "connection", required=True),
                _p("source_entity", "Source entity", "entity", required=True, connection_from="source_connection"),
                _p("target_connection", "Destination connection", "connection", required=True),
                _p("target_entity", "Destination entity", "entity", required=True, connection_from="target_connection"),
                _METHOD, _TRANSFORM]},
    {"slug": "file-import", "kind": "mapping", "icon": "bi-file-earmark-spreadsheet", "title": "Import a CSV / Excel file",
     "blurb": "Read rows from an uploaded file entity and push them into an API, matching columns by name.",
     "params": [_NAME,
                _p("source_connection", "File connection", "connection", required=True),
                _p("source_entity", "File entity", "entity", required=True, connection_from="source_connection", file_only=True),
                _p("target_connection", "Destination connection", "connection", required=True),
                _p("target_entity", "Destination entity", "entity", required=True, connection_from="target_connection"),
                _METHOD, _TRANSFORM]},
    {"slug": "fan-out", "kind": "mapping", "icon": "bi-diagram-3", "title": "Sync one source to several destinations",
     "blurb": "One source entity feeds any number of destination entities, each wired by matching field names.",
     "params": [_NAME,
                _p("source_connection", "Source connection", "connection", required=True),
                _p("source_entity", "Source entity", "entity", required=True, connection_from="source_connection"),
                _p("target_entities", "Destination entities", "entities", required=True,
                   help="Tick one or more — they can live on different connections."),
                _METHOD, _TRANSFORM]},

    # ── chains ──
    {"slug": "create-then-link", "kind": "chain", "icon": "bi-link-45deg", "title": "Create, then link",
     "blurb": "Create a record, capture its new id, then use that id in a second call (e.g. customer → order).",
     "params": [_CHAIN_NAME, _CONNECTION,
                _p("create_path", "Create — path", required=True, default="/customers"),
                _p("create_body", "Create — JSON body", "textarea", default='{"name": "Example"}'),
                _p("id_path", "Where the new id is in the response", default="id", help="Dotted JSON path, e.g. id or data.id"),
                _p("link_method", "Second call — method", "select", options=["POST", "PUT", "PATCH"], default="POST"),
                _p("link_path", "Second call — path", required=True, default="/orders"),
                _p("link_body", "Second call — JSON body", "textarea", default='{"customer_id": {{record_id}}}',
                   help="{{record_id}} is the id captured from the first call.")]},
    {"slug": "list-then-fetch", "kind": "chain", "icon": "bi-list-check", "title": "List, then fetch one",
     "blurb": "Read a list, capture the first item's id, and fetch that item's full record.",
     "params": [_CHAIN_NAME, _CONNECTION,
                _p("list_path", "List — path", required=True, default="/products"),
                _p("id_path", "Path to the first id in the list", default="0.id", help="0.id = first item; items.0.id if the list is wrapped."),
                _p("detail_path", "Detail — path", required=True, default="/products/{{first_id}}", help="{{first_id}} is the captured id.")]},
    {"slug": "async-export", "kind": "chain", "icon": "bi-hourglass-split", "title": "Start a job and wait for it",
     "blurb": "Kick off a long-running job (export, report, bulk import), poll until it finishes, then fetch the result.",
     "params": [_CHAIN_NAME, _CONNECTION,
                _p("start_method", "Start — method", "select", options=["POST", "GET"], default="POST"),
                _p("start_path", "Start — path", required=True, default="/exports"),
                _p("poll_path", "Status — path", required=True, default="/exports/{{start_job.id}}/status", help="{{start_job.<field>}} reads the start call's response."),
                _p("condition_path", "Status field", default="status"),
                _p("condition_value", "Finished when it equals", default="completed"),
                _p("result_path", "Result — path (optional)", default="/exports/{{start_job.id}}/download")]},

    # ── plans ──
    {"slug": "ordered-plan", "kind": "plan", "icon": "bi-list-ol", "title": "Ordered plan",
     "blurb": "Sequence existing mappings and chains — e.g. migrate, then verify — and run or schedule them together.",
     "params": [_p("name", "Name", required=True, placeholder="e.g. Nightly migration"),
                _p("steps", "Steps, in order", "steps", required=True)]},

    # ── runs (no server side: the Studio opens the mapping's run dialog) ──
    {"slug": "run-now", "kind": "run", "icon": "bi-play-circle", "title": "Run a migration now",
     "blurb": "Choose a mapping and start it immediately.", "action": "run",
     "params": [_p("mapping", "Mapping", "mapping", required=True)]},
    {"slug": "schedule-run", "kind": "run", "icon": "bi-clock", "title": "Schedule a migration",
     "blurb": "Choose a mapping, then set when it runs, a rate limit or input files.", "action": "schedule",
     "params": [_p("mapping", "Mapping", "mapping", required=True)]},
]


# ── Helpers ──────────────────────────────────────────────────────────────────

def _text(p, key, required=False, default=""):
    value = p.get(key)
    value = default if value is None else str(value).strip()
    if required and not value:
        raise TemplateError(f"“{key.replace('_', ' ')}” is required.")
    return value


def _get(model, pk, what):
    try:
        return model.objects.get(pk=int(pk))
    except (TypeError, ValueError, model.DoesNotExist):
        raise TemplateError(f"Choose a valid {what}.")


def _json_body(raw, what):
    """Same rule the chain editor applies: valid JSON once {{placeholders}} are stood in for."""
    if raw.strip():
        try:
            json.loads(re.sub(r"\{\{[^}]*\}\}", "null", raw))
        except ValueError:
            raise TemplateError(f"{what} must be valid JSON ({{{{placeholders}}}} aside).")
    return raw.strip()


def _norm(name):
    return re.sub(r"[^a-z0-9]", "", name.lower())


def auto_map(pair, preset="none"):
    """Wire every target field to the source field with the same (normalised) name —
    ignoring case, underscores and dashes, and falling back to the last segment of a
    dotted target name (precos.preco → preco). Returns (matched, unmatched target names)."""
    source_by_name = {}
    for f in pair.source_entity.fields.all():
        source_by_name.setdefault(_norm(f.name), f)
    matched, unmatched = 0, []
    for target in pair.target_entity.fields.all():
        source = source_by_name.get(_norm(target.name)) or source_by_name.get(_norm(target.name.split(".")[-1]))
        if not source:
            unmatched.append(target.name)
            continue
        text_to_text = source.field_type == Field.TYPE_STRING and target.field_type == Field.TYPE_STRING
        FieldMapping.objects.create(
            entity_mapping=pair, source_field=source, target_field=target,
            transform_rules=[{"op": preset}] if preset in ("trim", "uppercase", "lowercase") and text_to_text else [],
        )
        matched += 1
    return matched, unmatched


def _describe(pair, matched, unmatched):
    note = f"{pair.source_entity.name} → {pair.target_entity.name}: {matched} field(s) matched"
    if unmatched:
        shown = ", ".join(unmatched[:4]) + (f" +{len(unmatched) - 4} more" if len(unmatched) > 4 else "")
        note += f"; not matched: {shown}"
    return note + "."


# ── Appliers ─────────────────────────────────────────────────────────────────

def _write_method(p):
    method = _text(p, "write_method", default="POST").upper()
    if method not in ("POST", "PUT", "PATCH"):
        raise TemplateError("Choose POST, PUT or PATCH.")
    return method


@transaction.atomic
def apply_mapping(slug, p):
    name = _text(p, "name", required=True)
    source_conn = _get(Connection, p.get("source_connection"), "source connection")
    source = _get(Entity, p.get("source_entity"), "source entity")
    if source.connection_id != source_conn.id:
        raise TemplateError("The source entity doesn't belong to the source connection.")
    if slug == "file-import" and not source.source_file:
        raise TemplateError(f"“{source.name}” isn't backed by an uploaded file — add one from its connection's “From file” tab.")

    if slug == "fan-out":
        ids = p.get("target_entities") or []
        targets = [_get(Entity, i, "destination entity") for i in ids]
        if not targets:
            raise TemplateError("Tick at least one destination entity.")
    else:
        targets = [_get(Entity, p.get("target_entity"), "destination entity")]
        target_conn = _get(Connection, p.get("target_connection"), "destination connection")
        if targets[0].connection_id != target_conn.id:
            raise TemplateError("The destination entity doesn't belong to the destination connection.")
    if any(t.connection_id == source_conn.id for t in targets):
        raise TemplateError("A destination can't be on the source connection — pick a different system to write to.")

    mapping = Mapping.objects.create(name=name, source_connection=source_conn)
    mapping.destination_connections.add(*{t.connection for t in targets})
    notes = []
    for target in targets:
        pair = EntityMapping.objects.create(mapping=mapping, source_entity=source, target_entity=target, write_method=_write_method(p))
        matched, unmatched = auto_map(pair, _text(p, "transform", default="none"))
        notes.append(_describe(pair, matched, unmatched))
        if not target.endpoint_path:
            notes.append(f"“{target.name}” has no endpoint path yet — set one (edit entity) before running.")
    return {"kind": "mapping", "id": mapping.id, "notes": notes}


def _chain(p):
    name = _text(p, "name", required=True)
    return CallChain.objects.create(name=name, connection=_get(Connection, p.get("connection"), "connection"))


@transaction.atomic
def apply_chain(slug, p):
    chain = _chain(p)
    if slug == "create-then-link":
        CallChainStep.objects.create(
            chain=chain, order=1, name="create_record", method="POST", path=_text(p, "create_path", required=True),
            body=_json_body(_text(p, "create_body"), "The create body"),
            captures=[{"name": "record_id", "path": _text(p, "id_path", default="id")}])
        CallChainStep.objects.create(
            chain=chain, order=2, name="link_record", method=_text(p, "link_method", default="POST").upper(),
            path=_text(p, "link_path", required=True), body=_json_body(_text(p, "link_body"), "The second call's body"))
        notes = ["Step 1 creates the record and captures its id as {{record_id}}; step 2 uses it."]
    elif slug == "list-then-fetch":
        CallChainStep.objects.create(
            chain=chain, order=1, name="list_items", method="GET", path=_text(p, "list_path", required=True),
            captures=[{"name": "first_id", "path": _text(p, "id_path", default="0.id")}])
        CallChainStep.objects.create(chain=chain, order=2, name="get_item", method="GET", path=_text(p, "detail_path", required=True))
        notes = ["Step 1 lists and captures the first id as {{first_id}}; step 2 fetches that record."]
    elif slug == "async-export":
        result_path = _text(p, "result_path")
        CallChainStep.objects.create(
            chain=chain, order=1, name="start_job", method=_text(p, "start_method", default="POST").upper(),
            path=_text(p, "start_path", required=True), is_async=True,
            async_poll_path=_text(p, "poll_path", required=True), async_poll_method="GET",
            async_condition_path=_text(p, "condition_path", default="status"),
            async_condition_value=_text(p, "condition_value", default="completed"),
            async_interval_seconds=2.0, async_timeout_seconds=120.0, async_result_path=result_path)
        notes = ["One asynchronous step: it starts the job, polls until the status field matches, then reads the result."]
    else:
        raise TemplateError("Unknown chain template.")
    return {"kind": "chain", "id": chain.id, "notes": notes}


@transaction.atomic
def apply_plan(slug, p):
    name = _text(p, "name", required=True)
    steps = p.get("steps") or []
    if not steps:
        raise TemplateError("Add at least one step.")
    plan = MigrationPlan.objects.create(name=name, execution_mode=MigrationPlan.MODE_MIXED)
    for order, step in enumerate(steps, start=1):
        kind, pk = (step or {}).get("kind"), (step or {}).get("id")
        if kind == "chain":
            PlanStep.objects.create(plan=plan, order=order, chain=_get(CallChain, pk, "chain"))
        elif kind == "mapping":
            PlanStep.objects.create(plan=plan, order=order, mapping=_get(Mapping, pk, "mapping"))
        else:
            raise TemplateError("Each step must be a mapping or a chain.")
    return {"kind": "plan", "id": plan.id, "notes": [f"Draft plan with {len(steps)} step(s) — execute it from the toolbar when ready."]}


APPLY = {
    "copy-matching-fields": apply_mapping, "file-import": apply_mapping, "fan-out": apply_mapping,
    "create-then-link": apply_chain, "list-then-fetch": apply_chain, "async-export": apply_chain,
    "ordered-plan": apply_plan,
}


def apply_template(slug, params):
    if slug not in APPLY:
        raise TemplateError("Unknown template.")
    return APPLY[slug](slug, params or {})

File: datamigrator/static/js/studio/templates.js
/* Templates — starter recipes for the sidebar. The server (studio/templates.py)
   describes each one as metadata + a parameter schema; this file renders the
   gallery, builds the form from that schema, and hands the answers back to
   POST /studio/templates/<slug>/, which creates an ordinary mapping / chain / plan.

   Parameter types the form understands:
     text · textarea · select                      plain inputs
     connection                                    a connection picker
     entity      (+ connection_from, file_only)    an entity of the chosen connection
     entities                                      tick any number of entities, grouped by connection
     mapping                                       an existing mapping
     steps                                         an ordered list of mappings / chains (plans)
   Templates flagged `action` only *start* something: they open the chosen mapping
   and its run dialog rather than creating anything. */
(function () {
  'use strict';
  const S = window.Studio;
  const h = S.h;

  S.templates = [];
  const KIND_LABEL = { mapping: 'Migration', chain: 'Chain', plan: 'Plan', run: 'Run' };
  const KIND_ORDER = ['mapping', 'chain', 'plan', 'run'];
  const KIND_HELP = {
    mapping: 'Wire two systems together', chain: 'Script a sequence of API calls',
    plan: 'Run several things in order', run: 'Start something that already exists',
  };
  S.templateKindLabel = KIND_LABEL;

  S.loadTemplates = async function () {
    try { S.templates = (await S.api('/studio/templates/')).templates; } catch (e) { S.templates = []; }
    S.renderTree();
  };

  // ── Entities, fetched once per connection ────────────────────────────────
  const entityCache = new Map();
  const entitiesOf = async connectionId => {
    if (!entityCache.has(connectionId)) entityCache.set(connectionId, S.apiList(`/api/entities/?connection=${connectionId}`));
    return entityCache.get(connectionId);
  };
  // A template can be filled in right after the user discovers new entities elsewhere.
  S.forgetEntityCache = () => entityCache.clear();

  const opt = (value, text, selected) => `<option value="${S.esc(value)}" ${selected ? 'selected' : ''}>${S.esc(text)}</option>`;
  const placeholder = text => `<option value="">${S.esc(text)}</option>`;

  // ── Form builder ─────────────────────────────────────────────────────────
  // Returns { el, read() }: read() collects the answers (throws a user-facing Error when a
  // required one is missing). Fields that depend on another (entity ← connection) re-fill
  // themselves when it changes.
  function buildForm(template) {
    const controls = {};      // param name -> { el, read }
    const root = h('div', {});

    template.params.forEach(param => {
      const id = `tp_${param.name}`;
      let control;

      if (param.type === 'textarea') {
        const el = h('textarea', { class: 'form-control form-control-sm mono', id, rows: '3' });
        el.value = param.default || '';
        control = { el, read: () => el.value };

      } else if (param.type === 'select') {
        const opts = param.options.map(o => typeof o === 'string' ? { value: o, label: o } : o);
        const el = h('select', { class: 'form-select form-select-sm', id, html: opts.map(o => opt(o.value, o.label, o.value === param.default)).join('') });
        control = { el, read: () => el.value };

      } else if (param.type === 'connection') {
        const el = h('select', { class: 'form-select form-select-sm', id, html: placeholder('Choose a connection…') + S.tree.connections.map(c => opt(c.id, c.name)).join('') });
        control = { el, read: () => (el.value ? Number(el.value) : null) };

      } else if (param.type === 'mapping') {
        const el = h('select', { class: 'form-select form-select-sm', id, html: placeholder('Choose a mapping…') + S.tree.mappings.map(m => opt(m.id, m.name)).join('') });
        control = { el, read: () => (el.value ? Number(el.value) : null) };

      } else if (param.type === 'entity') {
        const el = h('select', { class: 'form-select form-select-sm', id, disabled: true, html: placeholder('Choose a connection first') });
        const fill = async () => {
          const parent = controls[param.connection_from];
          const connectionId = parent && parent.read();
          if (!connectionId) { el.innerHTML = placeholder('Choose a connection first'); el.disabled = true; return; }
          el.disabled = true;
          el.innerHTML = placeholder('Loading…');
          try {
            let list = await entitiesOf(connectionId);
            if (param.file_only) list = list.filter(e => e.has_file);
            el.innerHTML = list.length
              ? placeholder('Choose an entity…') + list.map(e => opt(e.id, `${e.name}  (${e.fields.length} fields)`)).join('')
              : placeholder(param.file_only ? 'No file-backed entities on this connection' : 'No entities discovered on this connection yet');
            el.disabled = !list.length;
          } catch (e) { el.innerHTML = placeholder('Could not load entities'); }
        };
        control = { el, read: () => (el.value ? Number(el.value) : null), fill };

      } else if (param.type === 'entities') {
        const el = h('div', { class: 'st-checklist', id }, h('div', { class: 'text-dim small', text: 'Loading entities…' }));
        (async () => {
          try {
            const groups = await Promise.all(S.tree.connections.map(async c => [c, await entitiesOf(c.id)]));
            const shown = groups.filter(([, list]) => list.length);
            el.replaceChildren(...(shown.length ? shown.flatMap(([c, list]) => [
              h('div', { class: 'st-check-group', text: c.name }),
              ...list.map(e => h('label', { class: 'd-flex gap-2 align-items-center' },
                h('input', { type: 'checkbox', class: 'form-check-input mt-0', value: e.id }), e.name, h('span', { class: 'text-dim small', text: `${e.fields.length} fields` }))),
            ]) : [h('div', { class: 'text-dim small', text: 'No entities discovered yet — discover some from a connection page.' })]));
          } catch (e) { el.textContent = 'Could not load entities.'; }
        })();
        control = { el, read: () => [...el.querySelectorAll('input:checked')].map(i => Number(i.value)) };

      } else if (param.type === 'steps') {
        const rows = h('div', {});
        const choices = [
          ...S.tree.mappings.map(m => ({ v: `mapping:${m.id}`, t: m.name, g: 'Mappings' })),
          ...S.tree.chains.map(c => ({ v: `chain:${c.id}`, t: c.name, g: 'Chains' })),
        ];
        const addRow = () => {
          const sel = h('select', { class: 'form-select form-select-sm',
            html: placeholder('Choose a mapping or chain…') + ['Mappings', 'Chains'].map(g => `<optgroup label="${g}">${choices.filter(c => c.g === g).map(c => opt(c.v, c.t)).join('')}</optgroup>`).join('') });
          const row = h('div', { class: 'st-dlg-row', dataset: { step: '1' } }, h('span', { class: 'st-step-n mono text-dim' }), sel,
            h('button', { type: 'button', class: 'btn btn-sm btn-outline-secondary', title: 'Move up', html: '&uarr;', onclick: () => { if (row.previousElementSibling) rows.insertBefore(row, row.previousElementSibling); number(); } }),
            h('button', { type: 'button', class: 'btn btn-sm btn-outline-secondary', title: 'Move down', html: '&darr;', onclick: () => { if (row.nextElementSibling) rows.insertBefore(row.nextElementSibling, row); number(); } }),
            h('button', { type: 'button', class: 'btn btn-sm btn-outline-danger', title: 'Remove', html: '&times;', onclick: () => { row.remove(); number(); } }));
          rows.append(row);
          number();
        };
        const number = () => [...rows.children].forEach((r, i) => { r.querySelector('.st-step-n').textContent = `${i + 1}.`; });
        const el = h('div', { id }, rows, h('button', { type: 'button', class: 'btn btn-sm btn-outline-secondary', text: '+ Add step', onclick: addRow }));
        if (!choices.length) rows.append(h('div', { class: 'text-dim small mb-1', text: 'No mappings or chains yet — create some first.' }));
        else addRow();
        control = {
          el,
          read: () => [...rows.querySelectorAll('[data-step]')].map(r => r.querySelector('select').value).filter(Boolean)
            .map(v => { const [kind, i] = v.split(':'); return { kind, id: Number(i) }; }),
        };

      } else {   // text
        const el = h('input', { class: 'form-control form-control-sm', id, placeholder: param.placeholder || '' });
        el.value = param.default || '';
        if (/path/.test(param.name)) el.classList.add('mono');
        control = { el, read: () => el.value.trim() };
      }

      controls[param.name] = control;
      root.append(h('div', { class: 'mb-3' },
        h('label', { class: 'form-label', for: id }, param.label, param.required ? h('span', { class: 'text-danger ms-1', text: '*' }) : null),
        control.el, param.help ? h('div', { class: 'form-text', text: param.help }) : null));
    });

    // Wire dependents (an entity list follows its connection picker).
    template.params.filter(p => p.type === 'entity').forEach(p => {
      const parent = controls[p.connection_from];
      if (parent) parent.el.addEventListener('change', () => controls[p.name].fill());
    });

    return {
      el: root,
      read() {
        const values = {};
        template.params.forEach(p => {
          const v = controls[p.name].read();
          const empty = v == null || v === '' || (Array.isArray(v) && !v.length);
          if (p.required && empty) throw new Error(`“${p.label}” is required.`);
          values[p.name] = v;
        });
        return values;
      },
    };
  }

  // ── The dialog for one template ──────────────────────────────────────────
  S.templateDialog = function (template) {
    const form = buildForm(template);
    const isAction = !!template.action;
    S.dialog({
      title: template.title, size: 'modal-lg',
      body: h('div', {},
        h('p', { class: 'text-dim small', text: template.blurb }),
        form.el,
        isAction ? null : h('div', { class: 'form-text', text: 'Everything created is an ordinary mapping, chain or plan — rename, rewire or delete it any time.' })),
      actions: [{ label: 'Cancel' }, {
        label: isAction ? (template.action === 'run' ? 'Run now' : 'Continue…') : 'Create', primary: true,
        onClick: async dlg => {
          let values;
          try { values = form.read(); } catch (e) { dlg.error(e.message); return false; }

          if (isAction) {
            const ed = await S.openDoc('mapping', values.mapping);
            if (ed) template.action === 'run' ? ed.run() : ed.runOptions();
            return;
          }
          const made = await S.api(`/studio/templates/${template.slug}/`, { method: 'POST', body: values });
          S.forgetEntityCache();
          await S.refreshTree();
          await S.openDoc(made.kind, made.id);
          S.toast(made.notes.join('  ·  '), 'ok', 7000);
        },
      }],
    });
  };

  // ── Gallery: every template, grouped ─────────────────────────────────────
  S.templateGallery = function () {
    const dlg = S.dialog({
      title: 'Start from a template', size: 'modal-lg',
      body: h('div', { class: 'st-tpl-gallery' }, KIND_ORDER.map(kind => {
        const items = S.templates.filter(t => t.kind === kind);
        if (!items.length) return null;
        return h('div', {},
          h('div', { class: 'st-tpl-group' }, h('b', { text: KIND_LABEL[kind] }), h('span', { text: KIND_HELP[kind] })),
          h('div', { class: 'st-tpl-grid' }, items.map(t => h('button', {
            type: 'button', class: 'st-tpl-card',
            onclick: () => { dlg.el.addEventListener('hidden.bs.modal', () => S.templateDialog(t), { once: true }); dlg.close(); },
          }, h('span', { class: 'st-tpl-icon' }, h('i', { class: t.icon })), h('span', { class: 'st-tpl-title', text: t.title }), h('span', { class: 'st-tpl-blurb', text: t.blurb })))));
      })),
    });
  };
})();
